src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+            0,
+            VM_PROT_READ | VM_PROT_WRITE | VM_PROT_COPY);
+    if (err == KERN_SUCCESS) {
+        auto typedAddr = reinterpret_cast<typename Hook::signature_t*>(addr);
+        *typedAddr = restore_original ? hook.d_original : intercept;
+        LOG_IF_DEBUG(symname << " intercepted!");
+    } else {
+        LOG(ERROR) << "Failed to patch " << symname;
+    }
+}
+
+static void
+patch_symbols_in_section(
+        const section_t* section,
+        uintptr_t slide,
+        const DynamicInfoTable& dyninfo_table,
+        bool restore_original)
+{
+    auto symbol_addr_table = reinterpret_cast<void**>(slide + section->addr);
+    for (unsigned int i = 0; i < section->size / sizeof(void*); i++) {
+        const char* symbol_name = dyninfo_table.getSymbol(section->reserved1, i);
+        if (!symbol_name || !symbol_name[0] || !symbol_name[1]) {
+            continue;
+        }
+#define FOR_EACH_HOOKED_FUNCTION(hookname)                                                              \
+    if (strcmp(hooks::hookname.d_symbol, symbol_name + 1) == 0) {                                       \
+        patch_symbol(                                                                                   \
+                hooks::hookname,                                                                        \
+                &intercept::hookname,                                                                   \
+                symbol_name,                                                                            \
+                symbol_addr_table + i,                                                                  \
+                section->size,                                                                          \
+                restore_original);                                                                      \
+        continue;                                                                                       \
+    }
+        MEMRAY_HOOKED_FUNCTIONS
+#undef FOR_EACH_HOOKED_FUNCTION
+    }
+}
+
+static void
+patch_symbols_in_shared_object(const struct mach_header* header, intptr_t slide, bool restore_original)
+{
+    Dl_info info;
+    if (dladdr(header, &info) == 0) {
+        return;
+    }
+
+    const char* image_name = info.dli_fname;
+    if (strstr(image_name, "memray.cpython") || strstr(image_name, "/dyld")
+        || strstr(image_name, "dyld.dylib"))
+    {
+        LOG_IF_DEBUG("Skipping patching symbols for " << image_name);
+        return;
+    }
+    LOG_IF_DEBUG("Patching symbols for " << image_name);
+
+    auto segment_filter = [](const char* seg_name) {
+        return strcmp(seg_name, SEG_DATA) == 0 || strcmp(seg_name, SEG_DATA_CONST) == 0
+               || strcmp(seg_name, SEG_AUTH_CONST) == 0;
+    };
+
+    DynamicInfoTable dyninfo_table(header, slide, segment_filter);
+    if (!dyninfo_table) {
+        LOG_IF_DEBUG("Could not construct dynamic information table" << image_name);
```

Comment:
Should this be an error, instead? Is this a condition that we expect to happen with some dynamic libraries, or would every occurrence of this indicate a problem with our MachO parsing?

If the latter, I think we'd want to complain loudly, so that users report problems they trigger to us.

Reply:
> s this a condition that we expect to happen with some dynamic libraries

Yeah, some system libraries do not have one of the 3 tables. This means that these libraries don't have PLT and we cannot patch them.